anvill/src/Optimize.cpp

```diff
@@ -84,63 +97,76 @@ void OptimizeModule(const EntityLifter &lifter_context,
-  mpm.run(module);
-
-  llvm::legacy::FunctionPassManager fpm(&module);
-  fpm.add(llvm::createDeadCodeEliminationPass());
-  fpm.add(llvm::createSinkingPass());
-  fpm.add(llvm::createNewGVNPass());
-  fpm.add(llvm::createSCCPPass());
-  fpm.add(llvm::createDeadStoreEliminationPass());
-  fpm.add(llvm::createSROAPass());
-  fpm.add(llvm::createEarlyCSEPass(true));
-  fpm.add(llvm::createBitTrackingDCEPass());
-  fpm.add(llvm::createCFGSimplificationPass());
-  fpm.add(llvm::createSinkingPass());
-  fpm.add(llvm::createCFGSimplificationPass());
-  fpm.add(llvm::createInstructionCombiningPass());
+  llvm::PassBuilder pass_builder;
+  llvm::LoopAnalysisManager lam;
+  llvm::FunctionAnalysisManager fam;
+  llvm::CGSCCAnalysisManager cgam;
+  llvm::ModuleAnalysisManager mam;
+  pass_builder.registerLoopAnalyses(lam);
+  pass_builder.registerFunctionAnalyses(fam);
+  pass_builder.registerCGSCCAnalyses(cgam);
+  pass_builder.registerModuleAnalyses(mam);
+  pass_builder.crossRegisterProxies(lam, fam, cgam, mam);
+
+  llvm::ModulePassManager mpm;
+  llvm::InlineParams inline_params;
+  inline_params.DefaultThreshold = 250;
+  mpm.addPass(llvm::ModuleInlinerWrapperPass(inline_params));
+  mpm.addPass(llvm::GlobalOptPass());
+  mpm.addPass(llvm::GlobalDCEPass());
+  // TODO(alex): Figure out what to do here.
+  // The new version of this pass only seems to have been added in LLVM 12.
+  // mpm.addPass(llvm::StripDeadDebugInfoPass());
+  mpm.run(module, mam);
+
+  llvm::FunctionPassManager fpm;
+  fpm.addPass(llvm::DCEPass());
+  fpm.addPass(llvm::SinkingPass());
+  fpm.addPass(llvm::NewGVNPass());
+  fpm.addPass(llvm::SCCPPass());
+  fpm.addPass(llvm::DSEPass());
+  fpm.addPass(llvm::SROA());
+  fpm.addPass(llvm::EarlyCSEPass());
+  fpm.addPass(llvm::BDCEPass());
+  fpm.addPass(llvm::SimplifyCFGPass());
+  fpm.addPass(llvm::SinkingPass());
+  fpm.addPass(llvm::SimplifyCFGPass());
+  fpm.addPass(llvm::InstCombinePass());
 
   auto error_manager_ptr = ITransformationErrorManager::Create();
   auto &err_man = *error_manager_ptr.get();
 
-  fpm.add(CreateSinkSelectionsIntoBranchTargets(err_man));
-  fpm.add(CreateRemoveUnusedFPClassificationCalls());
-  fpm.add(CreateRemoveDelaySlotIntrinsics());
-  fpm.add(CreateRemoveErrorIntrinsics());
-  fpm.add(CreateLowerRemillMemoryAccessIntrinsics());
-  fpm.add(CreateRemoveCompilerBarriers());
-  fpm.add(CreateLowerTypeHintIntrinsics());
-  fpm.add(CreateInstructionFolderPass(err_man));
-  fpm.add(llvm::createDeadCodeEliminationPass());
-  fpm.add(CreateRecoverEntityUseInformation(err_man, lifter_context));
-  fpm.add(CreateSinkSelectionsIntoBranchTargets(err_man));
-  fpm.add(CreateRemoveTrivialPhisAndSelects());
-  fpm.add(llvm::createDeadCodeEliminationPass());
-  fpm.add(CreateRecoverStackFrameInformation(err_man, options));
-  fpm.add(llvm::createSROAPass());
-  fpm.add(CreateSplitStackFrameAtReturnAddress(err_man));
-  fpm.add(llvm::createSROAPass());
+  AddSinkSelectionsIntoBranchTargets(fpm, err_man);
```

Comment:
The new pass manager doesn't take a pointer anymore. So if I want to do:
```
fpm.add(SinkSelectionsIntoBranchTargets(err_man)
```
Then I'd need to expose the headers for each of the passes and move them to `libraries/anvill_passes/include` since we need the actual definitions to construct it. To avoid reorganising everything, I've exposed these helpers in `Transform.h` to add the pass to the pass manager.